File: RemoveDuplicatesSortedArray.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Idea: Moving elements to the front if their current count is less than or equal to 2.
// We achieve this by using two pointers; one to traverse (fast); one to keep track of where
// the new element should be moved to (slow). Everything is up to the mark left of slow; so whenever
// we replace slow with an element; we move slow.

// Your code here along with comments explaining your approach

class RemoveDuplicatesSortedArray {
    public int removeDuplicates(int[] nums) {
        int fast = 0;
        int slow = 0;
        int count = 0;

        while (fast < nums.length) {
            if (fast == 0 || nums[fast] == nums[fast - 1]) {
                count++;
            } else {
                // encountered new element
                count = 1;
            }
            if (count <= 2) {
                // valid input; move it to the front
                nums[slow] = nums[fast];
                slow++;
            }
            fast++;
        }

        return slow;
    }
}

File: remove_duplicates_sortedarray.py
# Time complexity: O(n)
# Space complexity: O(1)
# Approach: initialise count as 0, start from index 1
# use two pointers i and j and start from index 1
# run the loop till i is at last index
# check element at index i and element before index i
# if same, increment count
# if count is less than or equal to 2, swap i and j elements

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        j = 1
        n = len(nums)
        count = 1
        if n <= 2:
            return n   
        for i in range(1,n):
            if nums[i]==nums[i-1]:
                count = count + 1
            else:
                count = 1
            if count <= 2:
                nums[j]=nums[i]
                j = j+1
        return j
            
        